clients/linux/src/background_work.rs

```diff
@@ -0,0 +1,120 @@
+use crate::messages::{Messenger, Msg};
+use std::sync::{Arc, Mutex};
+use std::thread;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+pub struct BackgroundWork {
+    pub messenger: Messenger,
+    pub auto_save_state: AutoSaveState,
+    pub auto_sync_state: AutoSyncState,
+}
+
+impl BackgroundWork {
+    pub fn default(m: &Messenger) -> Self {
+        Self {
+            messenger: m.clone(),
+            auto_save_state: AutoSaveState::default(),
+            auto_sync_state: AutoSyncState::default(),
+        }
+    }
+
+    pub fn current_time() -> u128 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_millis()
+    }
+
+    pub fn init_background_work(state: Arc<Mutex<Self>>) {
+        loop {
+            thread::sleep(Duration::from_secs(1));
+            let mut bgs = state.lock().unwrap();
+
+            if bgs.auto_save_state.is_active {
+                bgs.auto_save();
+            }
+
+            if bgs.auto_sync_state.is_active {
+                bgs.auto_sync();
+            }
+        }
+    }
+
+    pub fn auto_sync(&mut self) {
```

Comment:
I have a feeling you don't need `self.auto_sync_state.synced_last_edit` to exist.

Here's how I would probably model this function without that:

```
let unsaved_changes = last_change > last_save;
if unsaved_changes {
    return;
}

let unsynced_changes = last_save > last_sync;
let time_since_sync = ...
if unsynced_changes {
    if time_since_sync > Self::SYNC_AFTER_EDIT_DELAY {
        sync();
        return;
    } else {
        return;
    }
}

// There are no unsynced or unsaved changes at this point
if time_since_last_sync > TIME_BETWEEN_SYNCS {
    sync;
    return;
else {
    return;
}
```

You could declare all these helper variables up top and use if elses to avoid early returns, see which one is cleaner.